gpio: stm32: support disabling and reenabling interrupts on gpio pin

mniestroj opened the pull request:
Up to now interrupts could be only configured once, with no way to disable them in runtime.

Allow interrupts to be disabled in runtime and then properly reenabled on user request. This allows to ignore interrupts when software is not expecting them.

The improvement over previously reverted patch [1] is that we disable interrupts only when we configure port for which interrupt line was previously selected. This for example prevents to disable interrupts line 2 in case PA2 was previously configured as interrupt source, but we are currently configuring PB2 as output.

Besides we make a small improvement by declaring gpio_stm32_enable_int as static and splitting into smaller functions in order to achieve better readability.

Those patches are a result of previously noticed bug #19177.

[1] 0951ce2 ("gpio: stm32: support disabling and reenabling interrupts on pin")

mniestroj committed:
gpio: stm32: split helper functions from gpio_stm32_enable_int()

This patch doesn't change functionality, but is only related to improved
readability and reusability.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
gpio: stm32: statify gpio_stm32_enable_int()

This allows compiler to inline function body and reduce overall code
size.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
gpio: stm32: support disabling and reenabling interrupts on gpio pin

Up to now interrupts could be only configured once, with no way to
disable them in runtime.

Allow interrupts to be disabled in runtime and then properly reenabled
on user request. This allows to ignore interrupts when software is not
expecting them.

The improvement over previously reverted patch [1] is that we disable
interrupts only when we configure port for which interrupt line was
previously selected. This for example prevents to disable interrupts
line 2 in case PA2 was previously configured as interrupt source, but we
are currently configuring PB2 as output.

[1] 0951ce2d34b6 ("gpio: stm32: support disabling and reenabling
  interrupts on pin")

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

erwango commented:
@mniestroj, btw I didn't mentioned it previously but since this change is going to have much more impact than initially, you should be aware of imminent driver rework due to new GPIO API introduction, merged yesterday actually (https://github.com/zephyrproject-rtos/zephyr/pull/16648).
So please have a look and check if it fits your need.

mniestroj commented:
> So please have a look and check if it fits your need.

@erwango I don't see any problems, as those changes look unrelated. But thanks for pointing that, I'll try to follow it.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -185,10 +185,79 @@ int gpio_stm32_configure(u32_t *base_addr, int pin, int conf, int altf)
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	port = LL_GPIO_AF_GetEXTISource(line);
+#elif CONFIG_SOC_SERIES_STM32MP1X
```
`#elif defined(CONFIG_SOC_SERIES_STM32MP1X)`

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -185,10 +185,79 @@ int gpio_stm32_configure(u32_t *base_addr, int pin, int conf, int altf)
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	LL_GPIO_AF_SetEXTISource(port, line);
+#elif CONFIG_SOC_SERIES_STM32MP1X
```
`#elif defined(CONFIG_SOC_SERIES_STM32MP1X)`

erwango replied:
Then it would require a `#if defined(CONFIG_SOC_SERIES_STM32F1X)`

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -211,45 +280,25 @@ const int gpio_stm32_enable_int(int port, int pin)
+ */
+static int gpio_stm32_int_enabled_port(int pin)
+{
+	if (pin > 15) {
```
We could place this check directly in `gpio_stm32_get_exti_source` and remove this function. Having `gpio_stm32_int_enabled_port` doesn't seem to improve readability of the code.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -288,8 +337,11 @@ static int gpio_stm32_config(struct device *dev, int access_op,
+	if (flags & GPIO_INT) {
 		if (stm32_exti_set_callback(pin, cfg->port,
 					    gpio_stm32_isr, dev) != 0) {
 			return -EBUSY;
```
It doesn't quite belong to this PR, but also here we should have `goto release_lock;`. Maybe we should have an extra commit that cleans up this part of the code and ensures that hardware semaphore on dual core SoCs is always correctly released?

erwango replied:
@mnkp, indeed. I took care of this gpio driver update to new API, but would be cleaner if done here.

erwango replied:
I've pushed a dedicated PR for this purpose, you can rebase on top of it: https://github.com/zephyrproject-rtos/zephyr/pull/19646

mnkp reviewed (commented):
Maybe it would be better if this PR targeted `topic-gpio` branch?

erwango committed:
drivers/gpio: stm32: Rework configure function exit for dual core

With dual core handling introduction, we now need to take care to
always release lock before exiting function.
Rework gpio_stm32_config to take this into account.
Additionally, since ENOSYS usage is resevred to system calls
handling, replace with EIO.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -225,6 +225,35 @@ static void gpio_stm32_set_exti_source(int port, int pin)
+	port = LL_GPIO_AF_GetEXTISource(line);
+#elif CONFIG_SOC_SERIES_STM32MP1X
+	port = LL_EXTI_GetEXTISource(line);
+#elif defined(CONFIG_SOC_SERIES_STM32G0X)
```
Then for coherency,  use `#if defined(CONFIG_SOC_SERIES_STM32F1X)` if applying @mnkp's comment

erwango reviewed (changes requested):
@mniestroj, this PR is of particular interest for driver adaptation for GPIO API.
I've reused it on #19607 (and hence validated it as part of my tetsing).
So to make things a little bit more simple, would be nice to have it on gpio-topic branch.
Also, as mentioned by @mnkp, some changes on gpio_stm32_config function would be needed, but are not particularly linked to the purpose of this PR.
So, I've pushed https://github.com/zephyrproject-rtos/zephyr/pull/19646 on topic-gpio branch. It would be required that you base your change on that as well.

To sum up, after updating this PR wrt requested changes, can you rebase on top of #19646 and then push it on topic-gpio branch?

Thanks for your understanding on this unusualy complex PR handling.

mniestroj committed:
gpio: stm32: cleanup preprocessor macros

Check if CONFIG_SOC_SERIES_STM32MP1X is defined instead of its
value. While at it convert CONFIG_SOC_SERIES_STM32F1X check to be
consistent with others.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj commented:
@erwango I've changed code according to review comments and rebased on top of topic-gpio branch.

erwango commented:
@mniestroj, thanks!
I'm a little bit confused though, since I've finally done the changes myself directly in https://github.com/zephyrproject-rtos/zephyr/pull/19579 last week since we wanted to make it move fastly. I should have informed you to spare you the effort, sorry for this.
Anyway, your commits are present in #19579 and will be merged form this branch. 
Thanks for this contribution that has been useful to initiate the driver re-spawn.

mniestroj commented:
@erwango Ahh, I see that you fixed those issues in separate commit :) Thanks for reviewing and pulling those changes!

joerchan commented:
I think the reviewers list need to be cleaned up. Looks like many has been added by accident.

mniestroj commented:
Main changes from this PR has been merged part of #20443 PR.
